### PR TITLE
Br/dg/s3fault/eos 17129

### PR DESCRIPTION
**Pull request by DPG17**
Changes in common metadata for S3 fault tolerance feature

### CODE REVIEW

**Comment by gowthamchinna** [comment]
run mini-provisioning

File: server/s3_object_metadata.h [hunk 1]
```diff
@@ -131,6 +147,12 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
   // `true` in case of json parsing failure.
   bool json_parsing_error = false;
 
+  S3ObjectMetadataType obj_type;
+  // If multipart, PRTS is non-zero
+  unsigned int obj_parts;
+  unsigned int obj_fragments;
+  std::string pv_id;
```
Comment (knrajnambiar76) [inline server/s3_object_metadata.h:154]: pvid_str ?

File: server/s3_object_metadata.h [hunk 2]
```diff
@@ -232,7 +254,11 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
   virtual std::string get_upload_id();
   std::string& get_encoded_object_acl();
   std::string get_acl_as_xml();
-
+  unsigned int get_no_of_parts() { return this->obj_parts; }
```
Comment (knrajnambiar76) [inline server/s3_object_metadata.h:257]: get_number_of_parts seems better

Comment (knrajnambiar76) [inline server/s3_object_metadata.h:257]: get_number_of_parts seems better

File: server/s3_object_metadata.h [hunk 3]
```diff
@@ -347,5 +373,95 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
...
+// e.g., see object list table entry below:
+// ObjectOne                  <Basic object metadata> + versionID, FNo, PRTS​
+// ObjectTwo                  <Basic object metadata> + versionID, FNo, PRTS
+// ~ObjectOne|versionID|F1    OID1, PVID, size, layout-id1​
+// ~ObjectTwo|versionID|P1    OID2, PVID, size, layout-id​2
+// ~ObjectTwo|versionID|P2|F1 OID3, PVID, size, layout-id​3
+// ~ObjectTwo|versionID|P2|F2 OID4, PVID, size, layout-id​4
+class S3ObjectExtededMetadata : private S3ObjectMetadataCopyable {
+ private:
+  std::string bucket_name;
```
Comment (knrajnambiar76) [inline server/s3_object_metadata.h:397]: whether all these members needed here,or something derived from base class can be used

File: server/s3_object_metadata.h (hunk 1 above)
Comment (DPG17) [reply server/s3_object_metadata.h:154]: ok

File: server/s3_object_metadata.h (hunk 2 above)
Comment (DPG17) [reply server/s3_object_metadata.h:257]: ok

File: server/s3_object_metadata.h (hunk 3 above)
Comment (DPG17) [reply server/s3_object_metadata.h:397]: All members are needed. They are not in base class 'S3ObjectMetadataCopyable'

File: server/s3_factory.h [hunk 4]
```diff
@@ -79,6 +79,22 @@ class S3ObjectMetadataFactory {
     meta->set_object_list_index_oid(indx_oid);
     return meta;
   }
+  // Added below to create instance of 'S3ObjectExtededMetadata' class
+  virtual std::shared_ptr<S3ObjectExtededMetadata>
```
Comment (evgeniy-brazhnikov) [inline server/s3_factory.h:83]: ExteNded?

File: server/s3_factory.h [hunk 5]
```diff
@@ -79,6 +79,22 @@ class S3ObjectMetadataFactory {
...
+  create_object_ext_metadata_obj(std::shared_ptr<S3RequestObject> req,
+                                 const std::string& bucket_name,
+                                 const std::string& object_name,
+                                 const std::string& versionid,
+                                 unsigned int parts, unsigned int fragments,
```
Comment (evgeniy-brazhnikov) [inline server/s3_factory.h:88]: "int" is redundant. don't insist

File: server/s3_factory.h [hunk 6]
```diff
@@ -79,6 +79,22 @@ class S3ObjectMetadataFactory {
...
+                                 m0_uint128 indx_oid = {0ULL, 0ULL}) {
+    s3_log(S3_LOG_DEBUG, "",
+           "S3ObjectMetadataFactory::create_object_ext_metadata_obj\n");
+    std::shared_ptr<S3ObjectExtededMetadata> meta =
+        std::make_shared<S3ObjectExtededMetadata>(req, bucket_name, object_name,
```
Comment (evgeniy-brazhnikov) [inline server/s3_factory.h:93]: "req" isn't used below this point. so it's possible to use std::move(req)

File: server/s3_get_object_action.cc [hunk 7]
```diff
@@ -321,21 +321,32 @@ void S3GetObjectAction::check_full_or_range_object_read() {
...
-  read_object_data();
+  if (this->object_metadata->get_number_of_fragments() == 0) {
+    // Object is not fragmented
+    // get total number of blocks to read from an object
+    set_total_blocks_to_read_from_object();
+    motr_reader = motr_reader_factory->create_motr_reader(
+        request, object_metadata->get_oid(), object_metadata->get_layout_id());
+    // get the block,in which first_byte_offset_to_read is present
+    // and initilaize the last index with starting offset the block
+    size_t block_start_offset =
```
Comment (evgeniy-brazhnikov) [inline server/s3_get_object_action.cc:340]: Not sure that formula is correct. Can you please re-check?

File: server/s3_object_action_base.cc [hunk 8]
```diff
@@ -107,9 +107,31 @@ void S3ObjectAction::fetch_object_info() {
...
+        std::bind(&S3ObjectAction::fetch_ext_object_info_success, this),
+        std::bind(&S3ObjectAction::fetch_ext_object_info_failed, this));
+  } else {
+    next();
+  }
 }
 
+void S3ObjectAction::fetch_ext_object_info_success() { next(); }
+
+void S3ObjectAction::fetch_ext_object_info_failed() { next(); }
```
Comment (evgeniy-brazhnikov) [inline server/s3_object_action_base.cc:133]: Hmm... We have got an error. next()?

File: server/s3_object_action_base.h [hunk 9]
```diff
@@ -59,6 +60,8 @@ class S3ObjectAction : public S3Action {
   virtual void fetch_object_info_failed() = 0;
   virtual void fetch_object_info_success();
   virtual void fetch_bucket_info_success();
+  void fetch_ext_object_info_success();
```
Comment (evgeniy-brazhnikov) [inline server/s3_object_action_base.h:63]: Suggest add "virtual" keyword for future UTs

File: server/s3_object_metadata.h [hunk 10]
```diff
@@ -35,6 +35,15 @@
 #include "s3_request_object.h"
 #include "s3_timer.h"
 
+// Object list index table is divided into objects and their extended parts, if
+// any.
+// All extended object entries start/prefix with special character "~",
+// to make them at the bottom of object list index table.
+// This special character is chosen, as it is not allowed in S3 object name and
+// also higher in the lexographical order.
+#define EXTENDED_METADATA_OBJECT_PREFIX "~"
```
Comment (evgeniy-brazhnikov) [inline server/s3_object_metadata.h:44]: M.b. '~' ? i.e. single character instead of string (char array)

File: server/s3_object_metadata.h [hunk 11]
```diff
@@ -170,7 +192,7 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
                    std::shared_ptr<MotrAPI> motr_api = nullptr);
 
   S3ObjectMetadata(const S3ObjectMetadata&);
-
+  S3ObjectMetadataType get_object_type() { return this->obj_type; }
```
Comment (evgeniy-brazhnikov) [inline server/s3_object_metadata.h:195]: It's possible to add 'const' qualifier to member function definition.

File: server/s3_object_metadata.h [hunk 12]
```diff
@@ -232,7 +254,11 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
   virtual std::string get_upload_id();
   std::string& get_encoded_object_acl();
   std::string get_acl_as_xml();
-
+  unsigned int get_number_of_parts() { return this->obj_parts; }
```
Comment (evgeniy-brazhnikov) [inline server/s3_object_metadata.h:257]: All 3 member function can be defined as 'const'

File: server/s3_object_metadata.h [hunk 13]
```diff
@@ -232,7 +254,11 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
...
+  unsigned int get_number_of_fragments() { return this->obj_fragments; }
+  bool is_object_extended() {
+    return (obj_type != S3ObjectMetadataType::simple);
```
Comment (evgeniy-brazhnikov) [inline server/s3_object_metadata.h:260]: braces are redundant. dont insist

File: server/s3_object_metadata.h [hunk 14]
```diff
@@ -347,5 +373,95 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
...
+  std::shared_ptr<S3MotrKVSReader> motr_kv_reader;
+  std::shared_ptr<S3MotrKVSWriter> motr_kv_writer;
+  unsigned int fragments;
+  unsigned int parts;
+  S3ObjectMetadataState state;
+
+  // Key is: Multipart number (in case of fragments of multipart, e.g, 1, 2,
+  // etc), OR,
+  // 0 if fragments of simple object.
+  std::map<int, std::vector<struct s3_part_frag_context>> ext_objects;
```
Comment (evgeniy-brazhnikov) [inline server/s3_object_metadata.h:411]: FYI: std::map<> may re-arrange items

File: server/s3_object_metadata.h [hunk 15]
```diff
@@ -347,5 +373,95 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
...
+      std::shared_ptr<S3MotrKVSWriterFactory> kv_writer_factory = nullptr,
+      std::shared_ptr<MotrAPI> motr_api = nullptr);
+
+  // Used to report to caller.
+  std::function<void()> handler_on_success;
+  std::function<void()> handler_on_failed;
+
+  // Virtual Destructor
+  virtual ~S3ObjectExtededMetadata() {};
+
```
Comment (evgeniy-brazhnikov) [inline server/s3_object_metadata.h:437]: Are you sure that all member functions below need to be 'public'?

**Review by evgeniy-brazhnikov — APPROVED** [review]

**Review by knrajnambiar76 — APPROVED** [review]

**Pull request merged**